File: pidex/sources/sudo.py
import time

from pidex.core.constants import EVENT_SUDO_USED, SEVERITY_INFO, SOURCE_SUDO
from pidex.core.event import Event


def parse(entry: dict) -> Event | None:
    if entry.get("_COMM") != "sudo":
        return None

    message = entry.get("MESSAGE", "")
    parts = message.split(" ; ")
    if len(parts) < 4:
        return None

    user = parts[0].strip()
    command = parts[-1].replace("COMMAND=", "", 1).strip()

    return Event(
        source=SOURCE_SUDO,
        event_type=EVENT_SUDO_USED,
        severity=SEVERITY_INFO,
        title="Sudo Used",
        message=f"{user} ran sudo {command}",
        timestamp=_ts(entry),
    )


def _ts(entry: dict) -> float:
    return entry.get("__REALTIME_TIMESTAMP", time.time()) / 1_000_000

File: internal/parser/sudo.go
package parser

import (
	"regexp"

	"github.com/leadows/pi_dex/internal/core"
)

var sudoRE = regexp.MustCompile(`^\s*(\S+)\s*:.*COMMAND=(.*)`)

func ParseSudo(entry map[string]any) *core.Event {
	comm, _ := entry["_COMM"].(string)
	ident, _ := entry["SYSLOG_IDENTIFIER"].(string)
	if comm != "sudo" && ident != "sudo" {
		return nil
	}

	msg := entryMessage(entry)
	if msg == "" {
		return nil
	}

	m := sudoRE.FindStringSubmatch(msg)
	if m == nil {
		return nil
	}

	return &core.Event{
		Source:    core.SourcSudo,
		EventType: core.EventSudoUsed,
		Severity:  core.SeverityInfo,
		Title:     "Sudo Used",
		Message:   m[1] + " ran sudo " + m[2],
		Timestamp: entryTimestamp(entry),
	}
}
